### ISSUE TITLE
YANG 1.1 allows non-unique values in non-configuration leaf-lists

### ISSUE THREAD

**Issue by jabilmplane** [sysrepo/sysrepo#2379]
Hi.

Could you please help me to determine whether it is an issue or not.

Leaf-lists values used to be unique in YANG 1.0, but YANG 1.1 allows  non-unique values in non-configuration leaf-list.
RFC 6020 says: “The values in a leaf-list MUST be unique”
RFC 7950 has clarification: “In configuration data, the values in a leaf-list MUST be unique.”

I have prepared a simple test where get callback writes two identical values.
```
static int
dp_get_items_cb(sr_session_ctx_t *session, const char *module_name, const char *xpath, const char *request_xpath,
        uint32_t request_id, struct lyd_node **parent, void *private_data)
{
	*parent = lyd_new_path(NULL, sr_get_context(sr_session_get_connection(session)), "/test:root/frequeny-table", ( void* )"852", LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_UPDATE);
	lyd_new_path(*parent, NULL, "/test:root/frequeny-table", ( void* )"1052", LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_UPDATE);
	lyd_new_path(*parent, NULL, "/test:root/frequeny-table", ( void* )"7", LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_UPDATE);
	lyd_new_path(*parent, NULL, "/test:root/frequeny-table", ( void* )"0", LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_UPDATE);
	lyd_new_path(*parent, NULL, "/test:root/frequeny-table", ( void* )"7", LYD_ANYDATA_CONSTSTRING, LYD_PATH_OPT_UPDATE);

    return SR_ERR_OK;
}
```

I expect value 7 to be present in the output twice, but it is present only once:
```
> get --filter-xpath /test:root
DATA
<root xmlns="urn:test:1.0">
  <frequeny-table>852</frequeny-table>
  <frequeny-table>1052</frequeny-table>
  <frequeny-table>7</frequeny-table>
  <frequeny-table>0</frequeny-table>
</root>
```

Yang file is:
```
module test {
	yang-version 1.1;
	namespace "urn:test:1.0";
	prefix "test";

	container root
	{
		leaf-list frequeny-table
		{
			config false;
			type uint32;
		}
	}
}
```

Thanks in advance!

**Comment by jktjkt** [comment]
In my experience there is *some* support for this in sysrepo already, but when it comes to working with ops data that contain duplicate `leaf-list` values, you will soon hit troubles, even more with push-ops data because these are stored as diffs internally, and it's basically impossible to create a generic diff of a non-unique `leaf-list`. Some more details are at https://github.com/sysrepo/sysrepo/issues/2288#issuecomment-792379916 (and in the rest of that issue).

**Comment by michalvasko** [comment]
What @jktjkt said is true but to answer your question, why are you using `LYD_PATH_OPT_UPDATE` for all the calls of `lyd_new_path()`? With it you are essentially saying you do not want to create a node if it already exists so there you have your answer.

**Comment by jabilmplane** [comment]
So, which option would you recommend to use?

```
#define LYD_PATH_OPT_UPDATE   0x01 /**< If the target node exists, is a leaf, and it is updated with a new value or its
                                        default flag is changed, it is returned. If the target node exists and is not
                                        a leaf or generally no change occurs in the \p data_tree, NULL is returned and no error set. */
#define LYD_PATH_OPT_NOPARENT 0x02 /**< If any parents of the target node do not exist, return an error instead of implicitly
                                        creating them. */
#define LYD_PATH_OPT_OUTPUT   0x04 /**< Changes the behavior to ignoring RPC/action input schema nodes and using only output ones. */
#define LYD_PATH_OPT_DFLT     0x08 /**< The created node (nodes, if also creating the parents) is a default one. If working with
                                        data tree of type #LYD_OPT_DATA, #LYD_OPT_CONFIG, #LYD_OPT_RPC, #LYD_OPT_RPCREPLY, or
                                        #LYD_OPT_NOTIF, this flag is never needed and therefore should not be used. However, if
                                        the tree is #LYD_OPT_GET, #LYD_OPT_GETCONFIG, or #LYD_OPT_EDIT, the default nodes are not
                                        created during validation and using this flag one can set them (see @ref howtodatawd). */
#define LYD_PATH_OPT_NOPARENTRET 0x10 /**< Changes the return value in the way that even if some parents were created in
                                        addition to the path-referenced node, the path-referenced node will always be returned. */
#define LYD_PATH_OPT_EDIT     0x20 /**< Allows the creation of special leaves without value. These leaves are valid if used
                                        in a NETCONF edit-config with delete/remove operation. */
```

**Comment by michalvasko** [comment]
None of them, of course.

**Comment by jabilmplane** [comment]
Result is the same.
So, in fact, there is no way to see duplicates in the <get> output for leaf-list. Correct?

**Comment by jabilmplane** [comment]
@jktjkt
Thanks for the reply!
The real intent of this ticket is to figure out which behavior is correct. I do think that duplicates should be shown for "config false" leaf-list.
Another question whether it is possible to fix it in current implementation.

**Comment by jktjkt** [comment]
> The real intent of this ticket is to figure out which behavior is correct. I do think that duplicates should be shown for "config false" leaf-list.

Yes, that's true in YANG 1.1.

**Comment by jabilmplane** [comment]
@michalvasko, could you, please, share your viewpoint?

**Comment by michalvasko** [comment]
Considering this is currently an experimental feature at best, I do not want to spend much time debugging this. But I do not know why it is not working, there is a similar [test](https://github.com/sysrepo/sysrepo/blob/devel/tests/test_operational.c#L2335) with push operational data that works fine.

**Comment by Reshad-Rahman** [comment]
I am also running into this issue

**Comment by michalvasko** [comment]
Some fixes were implemented in *libyang* version 1.0.231 so try to update to that, if using an older version.
